File: go/format_test.go
/* Copyright (c) 2026 Richard Rodger, MIT License */


package aontu

import (
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

func TestFormatReportsWhatChanged(t *testing.T) {
	a := New()
	same := a.Format("a: 1\n")
	if "formatted" != same.Verdict || "a: 1\n" != same.Text || same.Changed {
		t.Fatalf("clean document: %+v", same)
	}

	crlf := a.Format("a: 1\r\n")
	if "a: 1\n" != crlf.Text || !crlf.Changed {
		t.Fatalf("crlf: %+v", crlf)
	}
}

// A document that does not parse is not formatted, and the report says
// why in the finding shape every verb uses, the file named.
func TestFormatRefusesASyntaxError(t *testing.T) {
	a := New()
	a.File = "broken.aon"
	r := a.Format("a: {b\n")
	if "error" != r.Verdict || 1 != len(r.Errors) {
		t.Fatalf("want one error: %+v", r)
	}
	f := r.Errors[0]
	if "syntax" != f.Code || "parse" != f.Class || "broken.aon" != f.Sites[0].File {
		t.Fatalf("finding: %+v", f)
	}

	// A merge-conflict marker is refused before the parse, as
	// everywhere else.
	m := New().Format("<<<<<<< HEAD\na: 1\n=======\na: 2\n>>>>>>> other\n")
	if "error" != m.Verdict || "merge_conflict" != m.Errors[0].Code {
		t.Fatalf("conflict: %+v", m)
	}
}

func TestFormatRefusesPastTheDepthBudget(t *testing.T) {
	nest := func(n int) string {
		return "a:" + strings.Repeat("{b:", n) + "1" + strings.Repeat("}", n) + "\n"
	}
	ok := New().Format(nest(999))
	if "formatted" != ok.Verdict || "a: "+strings.Repeat("b: ", 999)+"1\n" != ok.Text {
		t.Fatalf("999 levels: %s %q", ok.Verdict, ok.Text[:40])
	}
	deep := New().Format(nest(1000))
	if "error" != deep.Verdict || "max_depth" != deep.Errors[0].Code || "budget" != deep.Errors[0].Class {
		t.Fatalf("1000 levels: %+v", deep)
	}
}

func TestFormatRefusesItsOwnDefect(t *testing.T) {
	orig := formatSame
	defer func() { formatSame = orig }()

	formatSame = func(root Val, after string) bool { return false }
	r := New().Format("a: {b: 1}\n")
	if "error" != r.Verdict {
		t.Fatalf("want a refusal: %+v", r)
	}
	f := r.Errors[0]
	if "format_check" != f.Code || "internal" != f.Class ||
		`{"a":{"b":1}}` != *f.Expected || "a: b: 1\n" != *f.Actual ||
		"a formatter defect: please report it with the source" != *f.Note {
		t.Fatalf("finding: %+v", f)
	}

	a := New()
	a.File = "doc.aon"
	named := a.Format("a: 1\n")
	if "a formatter defect: please report it with the source (doc.aon)" != *named.Errors[0].Note {
		t.Fatalf("note: %q", *named.Errors[0].Note)
	}

	// The check sees the parsed root and the text about to be written.
	var seen []string
	formatSame = func(root Val, after string) bool {
		seen = []string{root.Canon(), after}
		return true
	}
	ok := New().Format("a: 1\n")
	if "formatted" != ok.Verdict || 2 != len(seen) || `{"a":1}` != seen[0] || "a: 1\n" != seen[1] {
		t.Fatalf("hook: %+v %v", ok, seen)
	}

	// And the real check, on a text that is not the document.
	if formatSameDocument(New().mustParse(t, "a: 1"), "a: 2\n") {
		t.Fatal("a different document passed the check")
	}
	if formatSameDocument(New().mustParse(t, "a: 1"), "a: {\n") {
		t.Fatal("an unclosed map is not `a: 1`")
	}
	if !formatSameDocument(New().mustParse(t, "a: {b: 1}"), "a: b: 1\n") {
		t.Fatal("a chain is the braced map")
	}
	if formatSameDocument(New().mustParse(t, "a: 1"), "a: {b\n") {
		t.Fatal("a text that does not parse is no document")
	}
}

func TestFormatKeepsTheSpellingTheEngineRefuses(t *testing.T) {
	orig := formatMeet
	defer func() { formatMeet = orig }()

	var seen [][2]string
	formatMeet = func(before, after string) bool {
		seen = append(seen, [2]string{before, after})
		return false
	}
	keep := New().Format("s: a: 1\ns: b: 2\n")
	if "formatted" != keep.Verdict || "s: a: 1\ns: b: 2\n" != keep.Text || keep.Changed {
		t.Fatalf("kept: %+v", keep)
	}
	if 1 != len(seen) || "s: a: 1\ns: b: 2\n" != seen[0][0] || "s: { a:1 b:2 }\n" != seen[0][1] {
		t.Fatalf("seen: %q", seen)
	}

	formatMeet = func(before, after string) bool { return true }
	if take := New().Format("s: a: 1\ns: b: 2\n"); "s: { a:1 b:2 }\n" != take.Text {
		t.Fatalf("taken: %q", take.Text)
	}

	// A statement inside a block is checked at its indentation, and
	// only it keeps its spelling before.
	W := strings.Repeat("x", 60)
	block := "s: {\n  a: 1\n  b: [\n    " + W + "\n    " + W + "\n  ]\n"
	seen = nil
	formatMeet = func(before, after string) bool {
		seen = append(seen, [2]string{before, after})
		return false
	}
	V := strings.Repeat("y", 70)
	inner := New().Format(block + "  c: { " + V + ": 1 d: 2 }\n}\n")
	if block+"  c: {\n    "+V+": 1\n    d: 2\n  }\n}\n" != inner.Text {
		t.Fatalf("inner: %q", inner.Text)
	}
	if 1 != len(seen) || "  c: {\n    "+V+": 1\n    d: 2\n  }\n" != seen[0][0] ||
		"  c: "+V+": 1\n  c: d: 2\n" != seen[0][1] {
		t.Fatalf("inner seen: %q", seen)
	}

	// The real check: the meet, and the kinds of failure, not their
	// count -- one unresolved reference is reported once by a block
	// and once per key it reaches by the statements.
	formatMeet = orig
	if !formatSameByMeet("s: a: $.t\ns: b: 2\n", "s: { a:$.t b:2 }\n") {
		t.Fatal("the same unresolved reference")
	}
	if formatSameByMeet("s: a: 1\n", "s: a: 2\n") {
		t.Fatal("a different value")
	}
	if !formatSameByMeet("c: { &: $.e }\nc: a: 1\nc: b: 2\n", "c: { &: $.e a:1 b:2 }\n") {
		t.Fatal("no_path once and no_path thrice are one kind")
	}

	// The outcome of generation is compared as GenerateVars decides it:
	// a nil root's own kind, and the relation verdict after a
	// generation that succeeded.
	if !formatSameByMeet("nil\n", "nil\n") || !strings.HasSuffix(formatMeetOf("nil\n"), "\nliteral_nil") {
		t.Fatalf("nil root: %q", formatMeetOf("nil\n"))
	}
	rel := "a: { dependsOn: rel() & inverse(usedBy) & [path($.b)] }\nb: {}\n"
	if !strings.HasSuffix(formatMeetOf(rel), "\nrelation_inverse_missing") {
		t.Fatalf("relation verdict: %q", formatMeetOf(rel))
	}

	raw, err := os.ReadFile(filepath.Join("..", "use-cases", "repros", "key-func", "spread-key-through-deep-ref.aon"))
	if err != nil {
		t.Fatal(err)
	}
	if taken := New().Format(string(raw)); !strings.Contains(taken.Text, "a: b: { c:d:e:$.a.b.f f: { &: { n:key() } x: {} } }\n") {
		t.Fatalf("repro: %q", taken.Text)
	}
}

func (a *Aontu) mustParse(t *testing.T, src string) Val {
	t.Helper()
	v, err := a.Parse(src)
	if err != nil {
		t.Fatal(err)
	}
	return v
}

// THE LINT (§4) is asked for, never assumed: without the option the
// report carries no findings, and with it the findings say where. The
// rules themselves are pinned row by row in fmt.tsv.
func TestFormatLintsOnlyWhenAsked(t *testing.T) {
	a := New()
	src := "a: 1\r\nHTTP_PORT: 8080\r\n"
	plain := a.Format(src)
	if nil == plain.Findings || 0 != len(plain.Findings) {
		t.Fatalf("unasked: %+v", plain.Findings)
	}
	linted := a.FormatWith(src, FormatOptions{Lint: true})
	// CRLF is read as LF before the lint counts, so the positions are
	// the file's lines.
	want := []LintFinding{{Rule: "style/key-case", Line: 2, Col: 1,
		Message: "key HTTP_PORT holds an underscore; httpPort would follow the form"}}
	if !reflect.DeepEqual(linted.Findings, want) || "a: 1\nHTTP_PORT: 8080\n" != linted.Text {
		t.Fatalf("linted: %+v %q", linted.Findings, linted.Text)
	}
	// A document that does not format has no findings to report.
	if bad := a.FormatWith("a: {b", FormatOptions{Lint: true}); "error" != bad.Verdict || nil != bad.Findings {
		t.Fatalf("error report: %+v", bad)
	}
}

func TestUnifiedDiff(t *testing.T) {
	cases := []struct{ name, before, after, want string }{
		{"same", "a\n", "a\n", ""},
		{"empty", "", "", ""},
		{"one line", "a\n", "b\n", "--- a/x\n+++ b/x\n@@ -1,1 +1,1 @@\n-a\n+b\n"},
		{"into empty", "", "a\nb\n", "--- a/x\n+++ b/x\n@@ -0,0 +1,2 @@\n+a\n+b\n"},
		{"to empty", "a\n", "", "--- a/x\n+++ b/x\n@@ -1,1 +0,0 @@\n-a\n"},
		// A missing final newline is a difference, and is said as diff
		// says it.
		{"no newline", "a", "a\n",
			"--- a/x\n+++ b/x\n@@ -1,1 +1,1 @@\n-a\n\\ No newline at end of file\n+a\n"},
		// Lines that repeat on both sides -- closers, blanks -- are no
		// anchors, and the gap between anchors recurses to the plain
		// delete-and-insert.
		{"repeated lines", "x: {\n  a: 1\n}\ny: {\n  b: 2\n}\n", "x: {\n  a: 1\n  c: 3\n}\ny: {\n  b: 2\n}\n",
			"--- a/x\n+++ b/x\n@@ -1,5 +1,6 @@\n x: {\n   a: 1\n+  c: 3\n }\n y: {\n   b: 2\n"},
		// Nothing in common at all: everything out, everything in.
		{"disjoint", "a\nb\n", "c\nd\n", "--- a/x\n+++ b/x\n@@ -1,2 +1,2 @@\n-a\n-b\n+c\n+d\n"},
		// A line that MOVED: the unique lines are out of order across
		// the sides, so the chain keeps one of them and the other is a
		// deletion and an insertion.
		{"moved line", "a\nb\nc\nd\n", "a\nc\nb\nd\n",
			"--- a/x\n+++ b/x\n@@ -1,4 +1,4 @@\n a\n-b\n c\n+b\n d\n"},
	}
	for _, c := range cases {
		if got := UnifiedDiff("x", c.before, c.after); got != c.want {
			t.Errorf("%s:\n want %q\n got  %q", c.name, c.want, got)
		}
	}

	lines := make([]string, 20)
	for i := range lines {
		lines[i] = "line " + itoa(i)
	}
	before := strings.Join(lines, "\n") + "\n"
	edited := append([]string{}, lines...)
	edited[2] = "changed 2"
	edited = append(edited[:17], append([]string{"inserted"}, edited[17:]...)...)
	want := "--- a/f.aon\n+++ b/f.aon\n" +
		"@@ -1,6 +1,6 @@\n line 0\n line 1\n-line 2\n+changed 2\n line 3\n line 4\n line 5\n" +
		"@@ -15,6 +15,7 @@\n line 14\n line 15\n line 16\n+inserted\n line 17\n line 18\n line 19\n"
	if got := UnifiedDiff("f.aon", before, strings.Join(edited, "\n")+"\n"); got != want {
		t.Fatalf("two hunks:\n want %q\n got  %q", want, got)
	}
}

// THE CORPUS GATE (FMT.0.md §7.5): every document in the repository
// that parses formats to a fixed point, and every one that does not is
// refused for its syntax, or its depth, and nothing else.
func TestEveryCorpusDocumentFormatsToAFixedPoint(t *testing.T) {
	var files []string
	for _, dir := range []string{filepath.Join("..", "use-cases"), filepath.Join("..", "test", "spec", "files")} {
		err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
			if err != nil {
				return err
			}
			if !info.IsDir() && strings.HasSuffix(path, ".aon") {
				files = append(files, path)
			}
			return nil
		})
		if err != nil {
			t.Fatal(err)
		}
	}
	if len(files) <= 300 {
		t.Fatalf("too few documents: %d", len(files))
	}
	var failures []string
	formatted := 0
	for _, file := range files {
		raw, err := os.ReadFile(file)
		if err != nil {
			t.Fatal(err)
		}
		a := New()
		a.File = file
		r := a.Format(string(raw))
		if "error" == r.Verdict {
			if "syntax" != r.Errors[0].Code && "max_depth" != r.Errors[0].Code {
				failures = append(failures, file+": refused with "+r.Errors[0].Code)
			}
			continue
		}
		formatted++
		again := a.Format(r.Text)
		switch {
		case "error" == again.Verdict:
			failures = append(failures, file+": the formatted text does not format: "+again.Errors[0].Code)
		case again.Text != r.Text:
			failures = append(failures, file+": not a fixed point")
		case again.Changed:
			failures = append(failures, file+": a fixed point that reports a change")
		}
	}
	if 0 < len(failures) {
		t.Fatalf("%s", strings.Join(failures, "\n"))
	}
	if formatted <= 300 {
		t.Fatalf("too few documents formatted: %d", formatted)
	}
}

func TestBundledModelsAreFormatted(t *testing.T) {
	a := New()
	names := aontuModels
	if 8 != len(names) ||
		"aontu:code" != names[0] ||
		"aontu:render" != names[1] ||
		"aontu:render/lang/go" != names[2] ||
		"aontu:render/lang/markdown" != names[3] ||
		"aontu:render/lang/text" != names[4] ||
		"aontu:render/lang/typescript" != names[5] ||
		"aontu:system" != names[6] ||
		"aontu:view" != names[7] {
		t.Fatalf("aontuModels: %v", names)
	}
	for _, name := range names {
		// `aontu fmt` normalises line endings, so a model embedded from a
		// CRLF checkout could never round-trip its own formatting.
		src := strings.ReplaceAll(aontuSources[name], "\r\n", "\n")
		rep := a.FormatWith(src, FormatOptions{Lint: true})
		if "formatted" != rep.Verdict {
			t.Fatalf("%s: verdict %s (%v)", name, rep.Verdict, rep.Errors)
		}
		if rep.Text != src {
			t.Fatalf("%s is not in the form aontu fmt writes", name)
		}
		if rep.Changed {
			t.Fatalf("%s: changed", name)
		}
		if 0 != len(rep.Findings) {
			t.Fatalf("%s has lint findings: %v", name, rep.Findings)
		}
	}
}

File: ts/test/format.test.ts
/* Copyright (c) 2026 Richard Rodger, MIT License */


import { describe, test } from 'node:test'
import * as Assert from 'node:assert'
import * as Fs from 'node:fs'
import * as Path from 'node:path'

import { format, unifiedDiff } from '../dist/aontu'
import { AONTU_SOURCES, AONTU_MODELS } from '../dist/aontumodel'


// The repository root, found from wherever the compiled test runs.
function repoRoot(): string {
  let dir = __dirname
  while (!Fs.existsSync(Path.join(dir, 'test', 'spec'))) {
    dir = Path.dirname(dir)
  }
  return dir
}

function aonFiles(dir: string, out: string[] = []): string[] {
  for (const name of Fs.readdirSync(dir).sort()) {
    const path = Path.join(dir, name)
    if (Fs.statSync(path).isDirectory()) {
      aonFiles(path, out)
    }
    else if (name.endsWith('.aon')) {
      out.push(path)
    }
  }
  return out
}


describe('format', () => {

  test('format-reports-what-changed', () => {
    const same: any = format('a: 1\n')
    Assert.equal(same.verdict, 'formatted')
    Assert.equal(same.text, 'a: 1\n')
    Assert.equal(same.changed, false)

    const crlf: any = format('a: 1\r\n')
    Assert.equal(crlf.text, 'a: 1\n')
    Assert.equal(crlf.changed, true)
  })

  // A document that does not parse is not formatted, and the report
  // says why in the finding shape every verb uses, the file named.
  test('format-refuses-a-syntax-error', () => {
    const r: any = format('a: {b\n', { path: 'broken.aon' })
    Assert.equal(r.verdict, 'error')
    Assert.equal(r.errors.length, 1)
    Assert.equal(r.errors[0].code, 'syntax')
    Assert.equal(r.errors[0].class, 'parse')
    Assert.equal(r.errors[0].sites[0].file, 'broken.aon')

    // A merge-conflict marker is refused before the parse, as
    // everywhere else.
    const m: any = format('<<<<<<< HEAD\na: 1\n=======\na: 2\n>>>>>>> other\n')
    Assert.equal(m.verdict, 'error')
    Assert.equal(m.errors[0].code, 'merge_conflict')
  })

  test('format-refuses-past-the-depth-budget', () => {
    const nest = (n: number) => 'a:' + '{b:'.repeat(n) + '1' + '}'.repeat(n) + '\n'
    const ok: any = format(nest(999))
    Assert.equal(ok.verdict, 'formatted')
    Assert.equal(ok.text, 'a: ' + 'b: '.repeat(999) + '1\n')
    const deep: any = format(nest(1000))
    Assert.equal(deep.verdict, 'error')
    Assert.equal(deep.errors[0].code, 'max_depth')
    Assert.equal(deep.errors[0].class, 'budget')
  })

  test('format-refuses-its-own-defect', () => {
    const r: any = format('a: {b: 1}\n', undefined, { same: () => false })
    Assert.equal(r.verdict, 'error')
    Assert.equal(r.errors[0].code, 'format_check')
    Assert.equal(r.errors[0].class, 'internal')
    Assert.equal(r.errors[0].expected, '{"a":{"b":1}}')
    Assert.equal(r.errors[0].actual, 'a: b: 1\n')
    Assert.equal(r.errors[0].note, 'a formatter defect: please report it with the source')

    const named: any = format('a: 1\n', { path: 'doc.aon' }, { same: () => false })
    Assert.equal(named.errors[0].note,
      'a formatter defect: please report it with the source (doc.aon)')

    // The hook sees the parsed root and the text about to be written.
    let seen: any
    const ok: any = format('a: 1\n', undefined, {
      same: (root, after) => ((seen = [root.canon, after]), true),
    })
    Assert.equal(ok.verdict, 'formatted')
    Assert.deepEqual(seen, ['{"a":1}', 'a: 1\n'])
  })

  test('format-keeps-the-spelling-the-engine-refuses', () => {
    const seen: string[][] = []
    const keep: any = format('s: a: 1\ns: b: 2\n', undefined, {
      meet: (before, after) => (seen.push([before, after]), false),
    })
    Assert.equal(keep.verdict, 'formatted')
    Assert.equal(keep.text, 's: a: 1\ns: b: 2\n')
    Assert.equal(keep.changed, false)
    Assert.deepEqual(seen, [['s: a: 1\ns: b: 2\n', 's: { a:1 b:2 }\n']])

    const take: any = format('s: a: 1\ns: b: 2\n', undefined, { meet: () => true })
    Assert.equal(take.text, 's: { a:1 b:2 }\n')

    // A statement inside a block is checked at its indentation, and
    // only it keeps its spelling before.
    const W = 'x'.repeat(60)
    const block = 's: {\n  a: 1\n  b: [\n    ' + W + '\n    ' + W + '\n  ]\n'
    seen.length = 0
    const V = 'y'.repeat(70)
    const inner: any = format(block + '  c: { ' + V + ': 1 d: 2 }\n}\n', undefined, {
      meet: (before, after) => (seen.push([before, after]), false),
    })
    Assert.equal(inner.text, block + '  c: {\n    ' + V + ': 1\n    d: 2\n  }\n}\n')
    Assert.deepEqual(seen, [[
      '  c: {\n    ' + V + ': 1\n    d: 2\n  }\n',
      '  c: ' + V + ': 1\n  c: d: 2\n',
    ]])

    const repro = Fs.readFileSync(Path.join(repoRoot(),
      'use-cases', 'repros', 'key-func', 'spread-key-through-deep-ref.aon'), 'utf8')
    const kept: any = format(repro)
    Assert.ok(kept.text.includes(
      'a: b: c: d: e: $.a.b.f\na: b: f: { &: { n:key() } }\na: b: f: x: {}\n'), kept.text)
  })

  test('format-lints-only-when-asked', () => {
    const src = 'a: 1\r\nHTTP_PORT: 8080\r\n'
    const plain: any = format(src)
    Assert.deepEqual(plain.findings, [])
    const linted: any = format(src, { lint: true })
    // CRLF is read as LF before the lint counts, so the positions are
    // the file's lines.
    Assert.deepEqual(linted.findings, [{
      rule: 'style/key-case', line: 2, col: 1,
      message: 'key HTTP_PORT holds an underscore; httpPort would follow the form',
    }])
    Assert.equal(linted.text, 'a: 1\nHTTP_PORT: 8080\n')
    // A document that does not format has no findings to report.
    Assert.equal('findings' in (format('a: {b', { lint: true }) as any), false)
  })

  test('unified-diff', () => {
    Assert.equal(unifiedDiff('x', 'a\n', 'a\n'), '')
    Assert.equal(unifiedDiff('x', '', ''), '')

    Assert.equal(unifiedDiff('x', 'a\n', 'b\n'),
      '--- a/x\n+++ b/x\n@@ -1,1 +1,1 @@\n-a\n+b\n')

    // Into an empty file, and out of one.
    Assert.equal(unifiedDiff('x', '', 'a\nb\n'),
      '--- a/x\n+++ b/x\n@@ -0,0 +1,2 @@\n+a\n+b\n')
    Assert.equal(unifiedDiff('x', 'a\n', ''),
      '--- a/x\n+++ b/x\n@@ -1,1 +0,0 @@\n-a\n')

    // A missing final newline is a difference, and is said as diff
    // says it.
    Assert.equal(unifiedDiff('x', 'a', 'a\n'),
      '--- a/x\n+++ b/x\n@@ -1,1 +1,1 @@\n-a\n\\ No newline at end of file\n+a\n')

    const lines = (n: number) => Array.from({ length: n }, (_, i) => 'line ' + i)
    const before = lines(20).join('\n') + '\n'
    const edited = lines(20)
    edited[2] = 'changed 2'
    edited.splice(17, 0, 'inserted')
    Assert.equal(unifiedDiff('f.aon', before, edited.join('\n') + '\n'),
      '--- a/f.aon\n+++ b/f.aon\n' +
      '@@ -1,6 +1,6 @@\n line 0\n line 1\n-line 2\n+changed 2\n line 3\n line 4\n line 5\n' +
      '@@ -15,6 +15,7 @@\n line 14\n line 15\n line 16\n+inserted\n line 17\n line 18\n line 19\n')

    const a = 'x: {\n  a: 1\n}\ny: {\n  b: 2\n}\n'
    const b = 'x: {\n  a: 1\n  c: 3\n}\ny: {\n  b: 2\n}\n'
    Assert.equal(unifiedDiff('x', a, b),
      '--- a/x\n+++ b/x\n@@ -1,5 +1,6 @@\n x: {\n   a: 1\n+  c: 3\n }\n y: {\n   b: 2\n')

    // Nothing in common at all: everything out, everything in.
    Assert.equal(unifiedDiff('x', 'a\nb\n', 'c\nd\n'),
      '--- a/x\n+++ b/x\n@@ -1,2 +1,2 @@\n-a\n-b\n+c\n+d\n')

    // A line that MOVED: the unique lines are out of order across the
    // sides, so the chain keeps one of them and the other is a deletion
    // and an insertion.
    Assert.equal(unifiedDiff('x', 'a\nb\nc\nd\n', 'a\nc\nb\nd\n'),
      '--- a/x\n+++ b/x\n@@ -1,4 +1,4 @@\n a\n-b\n c\n+b\n d\n')
  })

  // THE CORPUS GATE (FMT.0.md §7.5): every document in the repository
  // that parses formats to a fixed point, and every one that does not
  // is refused for its syntax, or its depth, and nothing else.
  test('every-corpus-document-formats-to-a-fixed-point', () => {
    const root = repoRoot()
    const files = [
      ...aonFiles(Path.join(root, 'use-cases')),
      ...aonFiles(Path.join(root, 'test', 'spec', 'files')),
    ]
    Assert.ok(300 < files.length, `too few documents: ${files.length}`)
    const failures: string[] = []
    let formatted = 0
    for (const file of files) {
      const src = Fs.readFileSync(file, 'utf8')
      const r: any = format(src, { path: file })
      const name = Path.relative(root, file)
      if ('error' === r.verdict) {
        if ('syntax' !== r.errors[0].code && 'max_depth' !== r.errors[0].code) {
          failures.push(`${name}: refused with ${r.errors[0].code}`)
        }
        continue
      }
      formatted++
      const again: any = format(r.text, { path: file })
      if ('error' === again.verdict) {
        failures.push(`${name}: the formatted text does not format: ${again.errors[0].code}`)
      }
      else if (again.text !== r.text) {
        failures.push(`${name}: not a fixed point`)
      }
      else if (again.changed) {
        failures.push(`${name}: a fixed point that reports a change`)
      }
    }
    Assert.deepEqual(failures, [], failures.join('\n'))
    Assert.ok(300 < formatted, `too few documents formatted: ${formatted}`)
  })
})


describe('format-bundled-models', () => {
  test('aontu-models-are-fmt-clean-and-lint-clean', () => {
    for (const name of AONTU_MODELS) {
      const src = AONTU_SOURCES[name]
      const report: any = format(src, { lint: true })
      Assert.equal(report.verdict, 'formatted', name)
      Assert.equal(report.text, src, name + ' is not in the form aontu fmt writes')
      Assert.equal(report.changed, false, name)
      Assert.deepEqual(report.findings, [], name + ' has lint findings')
    }
    Assert.deepEqual(AONTU_MODELS, ['aontu:code', 'aontu:lang/go',
      'aontu:lang/markdown', 'aontu:lang/text', 'aontu:lang/typescript',
      'aontu:profile', 'aontu:system', 'aontu:view'])
  })
})
